In [1]:
#Cargamos los datos para entener los datos
import pandas as pd 

#Usamos el enconding porque problablemente el dataset viene con caracteres que en UTF-8 no carga , evitamos error 
df = pd.read_csv('../data/spam.csv', encoding='latin-1', sep=';', names=['clase', 'mensajes'])
df.head(2)

clase                                           mensajes
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...

(HAM) = No Spam , (SPAM) = Spam

In [2]:
#Miramos cantidad de ejemplos que tenemos
df.shape

(5572, 2)

In [3]:
#Revisamos que los datos sean los correctos
df.dtypes

clase       object
mensajes    object
dtype: object

In [4]:
#Revisamos nulos
df.isnull().sum()

clase       0
mensajes    0
dtype: int64

In [5]:
#Revisamos balanceo en la variable clase
df['clase'].value_counts(normalize=True)

clase
ham     0.865937
spam    0.134063
Name: proportion, dtype: float64

In [6]:
#Revisamos duplicados 
df.duplicated().sum()

np.int64(404)

In [7]:
#Revisamos valores duplicados
df[df.duplicated()].head(10)

clase                                           mensajes
103   ham  As per your request 'Melle Melle (Oru Minnamin...
154   ham  As per your request 'Melle Melle (Oru Minnamin...
207   ham  As I entered my cabin my PA said, '' Happy B'd...
223   ham                             Sorry, I'll call later
326   ham                   No calls..messages..missed calls
339   ham                             Sorry, I'll call later
357  spam  Congratulations ur awarded 500 of CD vouchers ...
444   ham                             Sorry, I'll call later
533   ham                  Gudnite....tc...practice going on
655   ham                       Did u got that persons story

In [8]:
#Eliminamos los duplicados para evitar multicolonealidad
df.drop_duplicates(inplace=True)

In [9]:
#Rectificamos que ya no tengamos duplicados
df.duplicated().sum()

np.int64(0)

In [10]:
#Revisamos longitud de los para poder revisar cual de los textos es mas largo si los spam o los ham
df['longitud'] = df['mensajes'].apply(len)
df.groupby('clase')['longitud'].mean()

clase
ham      70.905004
spam    137.703988
Name: longitud, dtype: float64

Podemos ver como en promedio los que tienen spam son textos más largos 

In [ ]:
#Creamos un nuevo dataset , sin los datos duplicaods, con nuevas columnas lito para hacer los procedimientos de entrnamiento
df.to_csv('../data/spam_clean.csv', index=False)

### Conclusiones del análisis exploratorio

- Revisamos la **cantidad de ejemplos** disponibles para confirmar que contamos con una base de datos suficiente.  
- Verificamos que los datos estuvieran en el **tipo de dato correspondiente** y que no existieran valores nulos.  
- Analizamos el **balanceo de clases**, encontrando una distribución desigual con **13.4% de Spam** y **86.6% de Ham**.  
  - Esto nos indica un desbalance que probablemente tengamos que tratar más adelante, dependiendo de cómo se comporten las métricas (como **F1-score**) durante el entrenamiento.  
  - En ese caso, podríamos aplicar técnicas de balanceo como **oversampling** o **undersampling**.  
- Tratamos los **valores duplicados** para evitar multicolinealidad.  
- Revisamos la **longitud de los mensajes**, observando que en muchos casos los mensajes de spam tienden a ser más largos que los de ham.  
